// paystack-inline-ts/dist/index.js
import { PaystackPop as OriginalPaystackPop } from "./paystack";
var PaystackPop = /** @class */ (function () {
    function PaystackPop() {
        // initialize original PaystackPop
        var popup = new OriginalPaystackPop();
        this.pop = popup;
    }
    PaystackPop.prototype.isLoaded = function () {
        return this.pop.isLoaded();
    };
    PaystackPop.prototype.newTransaction = function (options) {
        return this.pop.newTransaction(options);
    };
    PaystackPop.prototype.resumeTransaction = function (options) {
        return this.pop.newTransaction(options);
    };
    PaystackPop.prototype.cancelTransaction = function (idOrTransaction) {
        return this.pop.cancelTransaction(idOrTransaction);
    };
    PaystackPop.prototype.preloadTransaction = function (options) {
        return this.pop.preloadTransaction(options);
    };
    PaystackPop.prototype.checkout = function (options) {
        return this.pop.checkout(options);
    };
    PaystackPop.prototype.paymentRequest = function (options) {
        return this.pop.paymentRequest(options);
    };
    return PaystackPop;
}());
export default PaystackPop;
